rspace/src/test/scala/coop/rchain/rspace/IStoreTests.scala

```diff
@@ -0,0 +1,158 @@
+  }
+
+  it should "remove obsolete channels" in withTestStore { store =>
+    forAll("channel", "datum") { (channel: String, datum: String) =>
+      val key  = List(channel)
+      val hash = store.hashChannels(key)
+      store.withTxn(store.createTxnWrite()) { txn =>
+        store.putDatum(txn, key, Datum(datum, persist = false))
+        // collectGarbage is called in removeDatum:
+        store.removeDatum(txn, key, 0)
+        store.getChannels(txn, hash) shouldBe empty
+      }
+      store.clear()
+    }
+  }
+
+  "putWaitingContinuation" should "put waiting continuation in a new channel" in withTestStore {
+    store =>
+      forAll("channel", "continuation") { (channel: String, pattern: String) =>
+        val key          = List(channel)
+        val patterns     = List(StringMatch(pattern))
+        val continuation = new StringsCaptor
+        val wc: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(patterns, continuation, false)
+
+        store.withTxn(store.createTxnWrite()) { txn =>
+          store.putWaitingContinuation(txn, key, wc)
+          store.getWaitingContinuation(txn, key) shouldBe List(wc)
+        }
+        store.clear()
+      }
+  }
+
+  it should "append continuation if channel already exists" in withTestStore { store =>
+    forAll("channel", "continuation") { (channel: String, pattern: String) =>
+      val key          = List(channel)
+      val patterns     = List(StringMatch(pattern))
+      val continuation = new StringsCaptor
+      val wc1: WaitingContinuation[Pattern, StringsCaptor] =
+        WaitingContinuation(patterns, continuation, false)
+
+      val wc2: WaitingContinuation[Pattern, StringsCaptor] =
+        WaitingContinuation(List(StringMatch(pattern + 2)), continuation, false)
+
+      store.withTxn(store.createTxnWrite()) { txn =>
+        store.putWaitingContinuation(txn, key, wc1)
+        store.putWaitingContinuation(txn, key, wc2)
+        store.getWaitingContinuation(txn, key) should contain theSameElementsAs List(wc1, wc2)
+      }
+      store.clear()
+    }
+  }
+
+  "removeWaitingContinuation" should "remove waiting continuation from index" ignore withTestStore {
+    store =>
+      forAll("channel", "continuation") { (channel: String, pattern: String) =>
+        val key          = List(channel)
+        val patterns     = List(StringMatch(pattern))
+        val continuation = new StringsCaptor
+        val wc1: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(patterns, continuation, false)
+        val wc2: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(List(StringMatch(pattern + 2)), continuation, false)
+
+        store.withTxn(store.createTxnWrite()) { txn =>
+          store.putWaitingContinuation(txn, key, wc1)
+          store.putWaitingContinuation(txn, key, wc2)
+          store.removeWaitingContinuation(txn, key, 0)
+          store.getWaitingContinuation(txn, key) shouldBe List(wc1)
```

Comment:
This test is ignored because implementations behave differently. One succeeds with `shouldBe List(wc1)` and other with `shouldBe List(wc2)`. Which one has a bug? @henrytill @ys-pyrofex

Reply:
Okay I think the correct strategy here is to treat `LMDBStore` as the correct implementation.  

The fix is fairly trivial:

```diff
diff --git a/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala b/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
index 1cd465e2..edbdf51d 100644
--- a/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
+++ b/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
@@ -84,7 +84,7 @@ class InMemoryStore[C, P, A, K <: Serializable] private (
     putCs(txn, channels)
     val waitingContinuations =
       _waitingContinuations.getOrElseUpdate(key, Seq.empty[WaitingContinuation[P, K]])
-    _waitingContinuations.update(key, waitingContinuations :+ continuation)
+    _waitingContinuations.update(key, continuation +: waitingContinuations)
   }
 
   private[rspace] def getData(txn: T, channels: Seq[C]): Seq[Datum[A]] =
```

Reply:
Corrected